### ISSUE TITLE
Hyper Parameter tuning with CV function in light gbm

### ISSUE THREAD

**Issue by Solly7** [microsoft/LightGBM#2862]
## Summary

I need help...I am tuning my parameters using cv funtion in light gbm. I have used num_boost_rounds=5000 and early_stopping_rounds=100 inside cv but it looks like the early_stopping_round is not activated becauase my model iterates up until 4999 in every iteration and my auc score stays the same. see below code and example of results.

#HyperParameter Tuning
gridsearch_params = [
    (max_depth, num_leaves)
    for max_depth in range(2,10,1)
    for num_leaves in range(30,120,1)
]

#initial best params and AUC
max_auc = float("Inf")
best_params = None
for max_depth, num_leaves in gridsearch_params:
    print("CV with max_depth={}, num_leaves={}".format(
                             max_depth,
                             num_leaves))
    # Update our parameters
    params['max_depth'] = max_depth
    params['num_leaves'] = num_leaves
    # Run CV
    cv_results = lgb.cv(
        params,
        lgtrain,
        num_boost_round=num_rounds,
        seed=42,
        nfold=5,
        metrics={'auc'},
        early_stopping_rounds=100,
        verbose_eval=100,
        stratified=False
    )
    # Update best AUC
    mean_auc = max(cv_results['auc-mean'])
    boost_rounds = np.argmax((cv_results['auc-mean']))
    print("\tAUC {} for {} rounds".format(mean_auc, boost_rounds))
    if mean_auc > max_auc:
        max_auc = mean_auc
        best_params = (max_depth,num_leaves)
print("Best params: {}, {}, AUC: {}".format(best_params[0], best_params[1], max_auc))

results
CV with max_depth=2, num_leaves=30
[4200]  cv_agg's auc: 0.741066 + 0.00236214
[4300]  cv_agg's auc: 0.741144 + 0.00231732
[4400]  cv_agg's auc: 0.7412 + 0.0023089
[4500]  cv_agg's auc: 0.74129 + 0.00233952
[4600]  cv_agg's auc: 0.741363 + 0.00233256
[4700]  cv_agg's auc: 0.741447 + 0.00233539
[4800]  cv_agg's auc: 0.741516 + 0.00233878
[4900]  cv_agg's auc: 0.741611 + 0.00231105
[5000]  cv_agg's auc: 0.741681 + 0.00228367
        AUC 0.7416807816736082 for 4999 rounds

CV with max_depth=2, num_leaves=31
[100]   cv_agg's auc: 0.704526 + 0.00201816
[200]   cv_agg's auc: 0.718909 + 0.00188427
[300]   cv_agg's auc: 0.723375 + 0.00190467
.....
.....
[4900]  cv_agg's auc: 0.741611 + 0.00231105
[5000]  cv_agg's auc: 0.741681 + 0.00228367
        AUC 0.7416807816736082 for 4999 rounds

**Comment by StrikerRUS** [comment]
@Solly7 
> it looks like the early_stopping_round is not activated becauase my model iterates up until 4999 in every iteration and my auc score stays the same

AUC metric should be maximized. According to the logs you have provided, AUC contiguously increases in your training process, e.g. 
```
[100] cv_agg's auc: 0.704526 + 0.00201816
[200] cv_agg's auc: 0.718909 + 0.00188427
[300] cv_agg's auc: 0.723375 + 0.00190467
.....
.....
[4900] cv_agg's auc: 0.741611 + 0.00231105
[5000] cv_agg's auc: 0.741681 + 0.00228367
```
So, everything is OK.

I'd recommend you using scikit-learn wrapper with GridSearchCV tool for the better experience.
